### ISSUE TITLE
[GLPI 9.4.4] Race condition in automatic actions code resulting in stall when using CLI actions.

### ISSUE THREAD

**Issue by callummoffat** [glpi-project/glpi#6564]
**Describe the bug**

Every week or so (timing can be between as little as three days and as much as a week), a race condition occurs which results in two automatic actions that use the CLI scheduler conflicting. On our system, we run the required Task Scheduler script every one minute. When the two actions stall, the entire automatic actions queue locks up, stopping ticket notifications.

**Page(s) URL**

If applicable, page(s) URL where the bug happens.

**To reproduce**

Set mailgate and queuednotification task to operate using the CLI scheduler.
Set up a Windows scheduled task recurring every **one minute** to trigger the scheduler

**Expected behavior**

Scheduled task runs without error, no queue lockup.

**Logs**

```
paste logs here
Find them in *-error.log files under glpi/files/_log/
```

**Screenshots**

**Your GLPI setup (you can find it in Setup > General menu, System tab)**

```
 
Operating system: Windows NT        10.0 build 14393 (Windows Server 2016) AMD64
PHP 7.3.5 cgi-fcgi (Core, PDO, Phar, Reflection, SPL, SimpleXML, Zend OPcache, bcmath, calendar, cgi-fcgi, ctype, curl, date,
	dom, exif, fileinfo, filter, gd, gettext, hash, iconv, imap, json, ldap, libxml, mbstring, mysqli, mysqlnd, openssl, pcre,
	pdo_mysql, pdo_sqlite, readline, session, soap, standard, tidy, tokenizer, wddx, xml, xmlreader, xmlrpc, xmlwriter, zip, zlib)
Setup: max_execution_time="300" memory_limit="128M" post_max_size="32M" safe_mode="" session.save_handler="files"
	upload_max_filesize="32M" 
Software: Microsoft-IIS/10.0
	Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/80.0.3956.0 Safari/537.36 Edg/80.0.328.0
Server Software: mariadb.org binary distribution
	Server Version: 10.3.18-MariaDB
	Server SQL Mode: 
	Parameters: @127.0.0.1/
	Host info: 127.0.0.1 via TCP/IP
	
mysqli extension is installed
ctype extension is installed
fileinfo extension is installed
json extension is installed
mbstring extension is installed
iconv extension is installed
zlib extension is installed
curl extension is installed
gd extension is installed
simplexml extension is installed
xml extension is installed
ldap extension is installed
imap extension is installed
Zend OPcache extension is installed
xmlrpc extension is installed
exif extension is installed
APCu extension is not present
CAS extension is not present
Database version seems correct (10.3.18) - Perfect!Database version seems correct (10.3.18) - Perfect!
OKD:\inetpub\wwwroot\GLPI/files/_log : OK
OKD:\inetpub\wwwroot\GLPI/config : OK
OKD:\inetpub\wwwroot\GLPI/files : OK
OKD:\inetpub\wwwroot\GLPI/files/_dumps : OK
OKD:\inetpub\wwwroot\GLPI/files/_sessions : OK
OKD:\inetpub\wwwroot\GLPI/files/_cron : OK
OKD:\inetpub\wwwroot\GLPI/files/_graphs : OK
OKD:\inetpub\wwwroot\GLPI/files/_lock : OK
OKD:\inetpub\wwwroot\GLPI/files/_plugins : OK
OKD:\inetpub\wwwroot\GLPI/files/_tmp : OK
OKD:\inetpub\wwwroot\GLPI/files/_cache : OK
OKD:\inetpub\wwwroot\GLPI/files/_rss : OK
OKD:\inetpub\wwwroot\GLPI/files/_uploads : OK
OKD:\inetpub\wwwroot\GLPI/files/_pictures : OK
Web access to files directory is protectedWeb access to files directory is protected : OK

```

**Additional context**

**Comment by cedric-anne** [comment]
Hi,

I am skeptical about the fact that the root cause is a race condition error. It is possible that the problem is caused by fatal error or something like that on one of the tasks process.

First, when it occurs, what is the status of each task (can be found in the form of corresponding task) ?
Second, could you check in log files if there is any error ?

Regards

**Comment by callummoffat** [comment]
Hi,

Will check in the log files for errors tomorrow, but status of task is stuck in 'running' whenever it stalls. Clearing the status unblocks the task and resumes normal functionality.

**Comment by cedric-anne** [comment]
Hi,

Which task is stuck in 'running' ? `mailgate` or `queuednotification` ?

If it is mailgate, then it may be related to a specific email. If you find the mail that blocks the task, can you attach it here as an eml file ?

Regards

**Comment by callummoffat** [comment]
Have had both or either lock up.

**Comment by cedric-anne** [comment]
Hi,

It is nearly impossible to investigate without any error trace. Could you try to run the cronjob with `--debug` argument and, if you reproduce the issue, give us content of your log files ?

Regards

**Comment by mulderr** [comment]
Run into quite similar issues. From time to time notifications stopped working. Happened around three times. Today again the `queuednotification` action was stuck in the running state and the notification queue was already filled up with hundreds of items.

I was manually sending notifications from the queue through Action > Send when I noticed a couple of the notifications at the top of the queue were always failing, giving no errors but a blank page after forcing send. The associated tickets had large attachments inside.

Turned out to be PHP memory_limit restriction which became obvious after turning on debug mode in account settings and retrying to send from the notification queue. The error was:

`PHP Fatal error: Allowed memory size of XXX bytes exhausted (tried to allocate YYY bytes) in /var/www/glpi/vendor/phpmailer/phpmailer/src/SMTP.php on line 656`

which was also logged into php www-error.log but alas not anywhere in `glpi/files/_log` which I've repeatedly checked before.

So in short, please try to check *all* the error logs including PHP and verify your memory_limit setting.

**Comment by trasher** [comment]
> `PHP Fatal error: Allowed memory size of XXX bytes exhausted (tried to allocate YYY bytes) in /var/www/glpi/vendor/phpmailer/phpmailer/src/SMTP.php on line 656`
> 
> which was also logged into php www-error.log but alas not anywhere in `glpi/files/_log` which I've repeatedly checked before.

Fatal PHP errrors are logged in system logs, this cannot be changed. When you have issues with a PHP application, you have to check both app logs (if that exists) and system logs.

**Comment by trasher** [comment]
Anyways, is there an apparent reason you had a memory issue? What was your memory_limit setting (you've hidden those informations in your error message.)?

**Comment by mulderr** [comment]
I thought the values may not be representative. The exact error was:
`PHP Fatal error: Allowed memory size of 134217728 bytes exhausted (tried to allocate 56611640 bytes) in /var/www/glpi/vendor/phpmailer/phpmailer/src/SMTP.php on line 656`

Memory limit was set to 128M by default. If the installation docs mention the need to change this then it's probably my fault to begin with :) I've now increased memory_limit to 1024M which may be a total overkill but we have spare RAM so I don't mind.

**Comment by trasher** [comment]
`Allowed memory size of 134217728 bytes exhausted` => this is the current memory_limit ;)

Indeed, increasing the memory limit is a good idea; I was just curious to know what kind of email would take so much resource at once :D

**Comment by cedric-anne** [comment]
For the memory issue, see #6573 .

For the initial issue, as we do not have logs, it is impossible to investigate.
